linera-chain/src/data_types.rs

```diff
@@ -697,11 +697,14 @@ impl ExecutedBlock {
     }
 
     pub fn required_blob_ids(&self) -> HashSet<BlobId> {
-        self.outcome.required_blob_ids()
+        let mut blob_ids = self.outcome.oracle_blob_ids();
+        blob_ids.extend(self.block.published_blob_ids());
+        blob_ids
     }
 
     pub fn requires_blob(&self, blob_id: &BlobId) -> bool {
-        self.required_blob_ids().contains(blob_id)
+        self.outcome.oracle_blob_ids().contains(blob_id)
+            || self.block.published_blob_ids().contains(blob_id)
```

Comment:
Will this method be called often? b/c both `oracle_blob_ids` and `published_blob_ids` are reconstructing the `BTreeSet` (no caching).

Reply:
Maybe they can become some kind of `LazyCell`s on `ExecutedBlock` which are set during construction and skipped when serializing.

Reply:
It's used only once in `download_certificate_for_blob`.

So memoizing wouldn't help at all, currently, and I don't think it's worth optimizing.

Reply:
But that in turn is used in `update_local_node_with_blobs_from` which iterates over `Vec<BlobId>`.

Reply:
Yes, but it receives only one certificate for every blob and then checks that the blob is indeed required by that certificate. So it's called only once on each certificate and we wouldn't gain anything with memoization.

Reply:
😵 right..